### ISSUE TITLE
Tensorflow saved model to Tensorflowjs model conversion 

### ISSUE THREAD

**Issue by purviljain** [tensorflow/tfjs#4212]
I have trained a model in Tensorflow and converting the model to Tensorflow.js

**I used the following command:**
`tensorflowjs_converter --control_flow_v2=True --input_format=tf_saved_model --saved_model_tags=serve --signature_name=serving_default --skip_op_check --strip_debug_ops=True --weight_shard_size_bytes=4194304 . tfjs`

My python model- [Link](https://drive.google.com/file/d/1dwOk3zuim3HT68LvAe_Q-pC311Ize7gy/view?usp=sharing)
Js model after conversion- [Link](https://drive.google.com/file/d/1Kl_CqkTyeX82dHiLSUGXSBY4nQz2pivR/view?usp=sharing)

**What do the warnings below mean? Do I lose accuracy of my model predictions on conversion?**

`WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_7_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_7_depthwise_BN/FusedBatchNormV3'`

**The full conversion log is:**

`(tfjs) tmk@vampire:~/Projects/WebDev/Uplara/SizingChart-TFJS-E2E/raw/pzero_model_2$ tensorflowjs_converter --control_flow_v2=True --input_format=tf_saved_model --saved_model_tags=serve --signature_name=serving_default --skip_op_check --strip_debug_ops=True --weight_shard_size_bytes=4194304 . tfjs
2020-11-11 01:55:44.509990: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-11-11 01:55:44.510022: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2020-11-11 01:55:46.329249: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2020-11-11 01:55:46.329275: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)
2020-11-11 01:55:46.329312: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (vampire): /proc/driver/nvidia/version does not exist
2020-11-11 01:55:46.329541: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-11-11 01:55:46.335162: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 3299990000 Hz
2020-11-11 01:55:46.335382: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55e3f1cfca40 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-11-11 01:55:46.335420: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-11-11 01:56:04.611672: I tensorflow/core/grappler/devices.cc:69] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0
2020-11-11 01:56:04.611802: I tensorflow/core/grappler/clusters/single_machine.cc:356] Starting new session
2020-11-11 01:56:04.765576: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:816] Optimization results for grappler item: graph_to_optimize
2020-11-11 01:56:04.765618: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   function_optimizer: Graph size after: 2828 nodes (2175), 2640 edges (1986), time = 79.521ms.
2020-11-11 01:56:04.765630: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   function_optimizer: function_optimizer did nothing. time = 1.276ms.
2020-11-11 01:56:07.116145: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:816] Optimization results for grappler item: graph_to_optimize
2020-11-11 01:56:07.116181: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   debug_stripper: debug_stripper did nothing. time = 0.725ms.
2020-11-11 01:56:07.116188: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   model_pruner: Graph size after: 1321 nodes (-1280), 1344 edges (-1281), time = 17ms.
2020-11-11 01:56:07.116193: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   constant_folding: Graph size after: 1319 nodes (-2), 1342 edges (-2), time = 98.689ms.
2020-11-11 01:56:07.116199: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   arithmetic_optimizer: Graph size after: 1319 nodes (0), 1342 edges (0), time = 27.223ms.
2020-11-11 01:56:07.116204: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   dependency_optimizer: Graph size after: 1315 nodes (-4), 1338 edges (-4), time = 12.049ms.
2020-11-11 01:56:07.116210: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   model_pruner: Graph size after: 1315 nodes (0), 1338 edges (0), time = 7.399ms.
2020-11-11 01:56:07.116215: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   constant_folding: Graph size after: 1315 nodes (0), 1338 edges (0), time = 32.807ms.
2020-11-11 01:56:07.116222: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   arithmetic_optimizer: Graph size after: 1315 nodes (0), 1338 edges (0), time = 27.599ms.
2020-11-11 01:56:07.116228: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   dependency_optimizer: Graph size after: 1315 nodes (0), 1338 edges (0), time = 13.628ms.
2020-11-11 01:56:07.116233: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   debug_stripper: debug_stripper did nothing. time = 1.454ms.
2020-11-11 01:56:07.116239: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   model_pruner: Graph size after: 1315 nodes (0), 1338 edges (0), time = 6.404ms.
2020-11-11 01:56:07.116245: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   constant_folding: Graph size after: 1315 nodes (0), 1338 edges (0), time = 29.492ms.
2020-11-11 01:56:07.116250: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   arithmetic_optimizer: Graph size after: 1315 nodes (0), 1338 edges (0), time = 26.115ms.
2020-11-11 01:56:07.116256: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   dependency_optimizer: Graph size after: 1315 nodes (0), 1338 edges (0), time = 13.382ms.
2020-11-11 01:56:07.116261: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   model_pruner: Graph size after: 1315 nodes (0), 1338 edges (0), time = 7.497ms.
2020-11-11 01:56:07.116266: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   constant_folding: Graph size after: 1315 nodes (0), 1338 edges (0), time = 29.918ms.
2020-11-11 01:56:07.116272: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   arithmetic_optimizer: Graph size after: 1315 nodes (0), 1338 edges (0), time = 26.138ms.
2020-11-11 01:56:07.116279: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   dependency_optimizer: Graph size after: 1315 nodes (0), 1338 edges (0), time = 14.685ms.
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_7_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_7_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_8_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_8_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_9_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_9_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_10_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_10_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_11_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_11_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_12_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_12_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_13_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_13_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_14_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_14_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_15_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_15_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_11/deep_labv3lite/my_mobile_net/functional_9/block_16_depthwise_BN/FusedBatchNormV3'
WARNING:tensorflow:Didn't find expected Conv2D or DepthwiseConv2dNative input to 'StatefulPartitionedCall/functional_17/functional_15/deep_labv3/my_mobile_net_1/functional_13/block_16_depthwise_BN/FusedBatchNormV3'
2020-11-11 01:56:20.055959: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:816] Optimization results for grappler item: graph_to_optimize
2020-11-11 01:56:20.055989: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   remapper: Graph size after: 1397 nodes (2), 1030 edges (22), time = 22.547ms.
2020-11-11 01:56:20.055995: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   constant_folding: Graph size after: 827 nodes (-570), 850 edges (-180), time = 53.079ms.
2020-11-11 01:56:20.056000: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   arithmetic_optimizer: Graph size after: 827 nodes (0), 850 edges (0), time = 20.648ms.
2020-11-11 01:56:20.056004: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   dependency_optimizer: Graph size after: 827 nodes (0), 850 edges (0), time = 12.759ms.
2020-11-11 01:56:20.056008: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   remapper: Graph size after: 827 nodes (0), 850 edges (0), time = 4.52ms.
2020-11-11 01:56:20.056013: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   constant_folding: Graph size after: 827 nodes (0), 850 edges (0), time = 22.811ms.
2020-11-11 01:56:20.056018: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   arithmetic_optimizer: Graph size after: 827 nodes (0), 850 edges (0), time = 19.165ms.
2020-11-11 01:56:20.056024: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   dependency_optimizer: Graph size after: 827 nodes (0), 850 edges (0), time = 8.309ms.
Writing weight file tfjs/model.json...`

### **TLDR**
What is the best way to convert my tensorflow saved model without losing any accuracy?

**Comment by pyu10055** [comment]
This warning message is not related to the model accuracy, it is for indicating the pattern matching failure during BatchNorm op fusing into conv2d. 

Inference result divergency can be caused by op kernel implementation differences between TF and TFJS, or due to the browser texture precision.
Can you specify OS/browser/TFJS versions etc? thanks

**Comment by purviljain** [comment]
I am running this in Nodejs.

OS version: Ubuntu 20.04
TFJS version: 2.1.0

Also, how should I figure out which operations have kernel implementation differences between TF and TFJS? And if there are, do I have to skip those operations or is there some other way to get exactly same outputs?

**Comment by purviljain** [comment]
Is it better if I sequentially run 5 TFJS models one after the other, or would it be better if I stitch them all into one graph? 

WIll there be a performance difference between these two? If yes, which would be better?

**Comment by pyu10055** [comment]
@purviljain Are you saying you have multiple models, and are running multiple models back to back? If it would be better to evaluate each model (comparing python and TFJS outputs given the same input), it would help you to isolate the problem.

**Comment by abbhinavvenkat** [comment]
@pyu10055 
a. We feel it is because of some operation in TfJs being incompatible, or, like you pointed out, possibly with different implementation in Tf and TfJs. Do you know of a way to check such operations? 
b. We've tried a single model: Resnet based Encoder-Decoder as well (https://github.com/tensorflow/tfjs/issues/4214) and notice a drop in accuracy. How do you suggest we proceed?

**Comment by purviljain** [comment]
Hey @rthadur , our issue hasn't been resolved yet
It would be great if we could get some more insights on the two problems stated in the previous comment.
Thank you!

**Comment by pyu10055** [comment]
@abbhinavvenkat @purviljain One way to debug the difference is walk through the model topology node by node, and compare tensor values between python and TFJS. This would likely identify the problematic op. TensorFlow has a tool called TFDbg where you can show tensor values within the session, and in TFJS you need to put breakpoint in the ModelExecutor class to iterate through the op execution.

If you can share the model, we might be able to help you on that.

**Comment by purviljain** [comment]
Hey, sorry for the delay, we're working on a smaller model so it is easier for us to debug. We will share that shortly.
Can you also share your email-id with us? We would really prefer if we could share our model privately with you.
Thank you!

**Comment by pyu10055** [comment]
@purviljain You can create a private github repo and invite my github handle to the repo.

**Comment by purviljain** [comment]
Hey we have created the private repo and sent you an invite. 
The repo contains the python model file, the TFJS code, few sample inputs and outputs illustrating the difference in outputs between python and TFJS models.
Please let us know if anything else is required.
Really appreciate your help on this!

**Comment by rohanmuplara** [comment]
hey @pyu10055 I think we have identified the problem. The problem is in the upsampling2d layer when we are using bilinear. The problem is the tensorflow python code and the tensorflowjs code use different methods for bilinear interpolation. The tensorflow code uses the half-pixel method while the tfjs code uses two different methods depending on align corners and neither one is similar. 

python version - https://colab.research.google.com/drive/1ID3yDkG8vyODerZbaAOeK5QS_06v8A8v?usp=sharing
javascript version - https://codepen.io/rohan-uplara/pen/mdrEeMw; look at console
different approaches -  (https://jricheimer.github.io/tensorflow/2019/02/11/resize-confusion/ )  What would you suggest for us to do? 
Specific Question 
1. Is resizing supported? https://js.tensorflow.org/api/latest/#image.resizeBilinear is not there but in the master https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/ops/image/resize_bilinear.ts#L47 it is. Is it possible somehow to use the master and if so how do build it? Second, if we are able to get this custom layer to work, how do we make sure when we do the conversion to make sure it calls resizeBilinear in the right way. Right now by default tensorflow has half-pixels enabled and no way to turn it off  and tensorflowjs has it false. 

Any help would be appreciated.

**Comment by pyu10055** [comment]
@rohanmuplara @purviljain we are adding support for bilinear resize in the UpSampling2D layer, please refer to the issue #4386

**Comment by rohanmuplara** [comment]
Thanks so much @pyu10055 !
